x509-identity-mgmt/js/src/services/CertificateService.js

```diff
@@ -0,0 +1,309 @@
+          throw this.error.BadRequest('The fingerprint of the CA defined in the payload of the request '
+            + 'does not match the fingerprint of the certificate of the root CA that was '
+            + 'also informed in the certificate chain.');
+        }
+      }
+    }
+
+    if (!rootCAFingerprint) {
+      throw this.error.BadRequest('It was not possible to obtain the fingerprint of the root CA. '
+        + 'Make sure to inform it in the request payload or inform the root CA '
+        + 'certificate as the last one in the chain of certificates.');
+    }
+
+    // Attempts to obtain the previously registered root CA certificate...
+    let registeredRootCAPem = null;
+    try {
+      registeredRootCAPem = await this.trustedCAService.getPEM(rootCAFingerprint);
+    } catch (ex) {
+      this.logger.debug(ex);
+      // If the root CA certificate has not been registered, but it was informed in
+      // the request and the service is able to automatically register it, then it does.
+      // Otherwise, it throws an error and does not continue with the operation.
+      if (this.caCertAutoRegistration && rootCAPem) {
+        await this.trustedCAService.registerCertificate({
+          caPem: rootCAPem, allowAutoRegistration: false,
+        });
+        registeredRootCAPem = rootCAPem;
+      } else {
+        throw this.error.BadRequest('The root CA certificate is not registered as a trusted certificate yet.');
+      }
+    }
+
+    // The last certificate of the chain must be that of the root CA previously registered
+    certificateChain.push(registeredRootCAPem);
+
+    // Converts the PEM format certificate chain to javascript objects
+    const certChain = certificateChain.map((pem) => this.pkiUtils.parseCert(pem));
+
+    // ensures that the certificate chain of trust is valid
+    const trusted = certChain[certChain.length - 1];
+    await this.pkiUtils.checkChainOfTrust(certChain, trusted);
+
+    // Register the certificate in the database
+    const subjectDN = this.dnUtils.from(certToBeRegistered.subject).stringify();
+    const model = new this.CertificateModel({
+      caFingerprint: rootCAFingerprint,
+      fingerprint: certFingerprint,
+      pem: pemToBeRegistered,
+      subjectDN,
+      validity: {
+        notBefore: certToBeRegistered.notBefore.value,
+        notAfter: certToBeRegistered.notAfter.value,
+      },
+      issuedByDojotPki: false,
+      belongsTo,
+      tenant: this.tenant,
+    });
+    await model.save();
+
+    return { certificateFingerprint: certFingerprint };
+  }
+
+  /**
+   * Changes the ownership that is associated with a certificate
+   *
+   * @param {object} filterFields Filter fields to find the correct certificate in the database
+   * @param {object} belongsTo Data of whom the certificate should be associated
+   *
+   * @throws an exception if no record is found with the entered filters.
+   */
+  async changeOwnership(filterFields, belongsTo) {
```

Comment:
If we consider that the cname in the CSR refers to the device to which the certificate belongs to, we need some caution here regarding certificates signed by dojot.

Reply:
With de IoTAgent relying on the certificate's fingerprint, we no longer have to worry about any value of the `Subject DN` fields. I think the association should be only in the `belongsTo` attribute of the request's payload.

Reply:
OK!